File: entities/Choice.py
import random

class Choice:
    def __init__(self, number, choices):
        self.number = number
        self.choices = choices

    def getRandomChoice(self, items = None):
        if items is not None:
            filtered_choices = [choice for choice in self.choices if choice not in items]
        else:
            filtered_choices = self.choices

        if len(filtered_choices) <= self.number:
            return filtered_choices

        return random.sample(filtered_choices, self.number)

    def getRandomChoiceWithoutCheck(self):
        if len(self.choices) <= self.number:
            return self.choices

        return random.sample(self.choices, self.number)

    def __str__(self):
        return f"{self.number} - {', '.join(str(choice) for choice in self.choices)}"

File: TrainDataGen/Entities/Choice.cs
namespace TrainDataGen.Entities;

public class Choice
{
    public int Number { get; set; }
    public List<string> Choices { get; set; }

    public Choice(int number, List<string> choices)
    {
        Number = number;
        Choices = choices;
    }

    public List<string> GetRandomChoice(IEnumerable<string>? items = null)
    {
        var filteredChoices = items != null
            ? Choices.Where(choice => items == null || !items.Contains(choice)).ToList()
            : Choices;

        if (filteredChoices.Count <= Number)
            return filteredChoices;

        return GetRandomSample(filteredChoices, Number);
    }

    public List<string> GetRandomChoiceWithoutCheck()
    {
        if (Choices.Count <= Number)
            return Choices;

        return GetRandomSample(Choices, Number);
    }

    public override string ToString()
    {
        return $"{Number} - {string.Join(", ", Choices)}";
    }

    private static List<string> GetRandomSample(List<string> source, int count)
    {
        var rng = new Random();
        return source.OrderBy(_ => rng.Next()).Take(count).ToList();
    }
}